pageserver/src/page_service.rs

```diff
@@ -145,6 +164,180 @@ struct FeQueryMessage {
     body: Bytes,
 }
 
+// We only support the simple case of Parse on unnamed prepared statement and
+// no params
+#[derive(Debug)]
+struct FeParseMessage {
+    query_string: Bytes,
+}
+
+fn read_null_terminated(buf: &mut Bytes) -> Result<Bytes>
+{
+    let mut result = BytesMut::new();
+
+    loop {
+        if !buf.has_remaining() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "no null-terminator in string",
+            ));
+        }
+
+        let byte = buf.get_u8();
+
+        if byte == 0 {
+            break;
+        }
+        result.put_u8(byte);
+    }
+    return Ok(result.freeze());
+}
+
+impl FeParseMessage {
+    pub fn parse(body: Bytes) -> Result<FeMessage> {
+        let mut buf = body.clone();
+        let _pstmt_name = read_null_terminated(&mut buf)?;
+        let query_string = read_null_terminated(&mut buf)?;
+        let nparams = buf.get_i16();
+
+        // FIXME: the rust-postgres driver uses a named prepared statement
+        // for copy_out(). We're not prepared to handle that correctly. For
+        // now, just ignore the statement name, assuming that the client never
+        // uses more than one prepared statement at a time.
+        /*
+        if pstmt_name.len() != 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "named prepared statements not implemented in Parse",
+            ));
+        }
+         */
+
+        if nparams != 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "query params not implemented",
+            ));
+        }
+
+
+        Ok(FeMessage::Parse(FeParseMessage {query_string}))
+    }
+}
+
+#[derive(Debug)]
+struct FeDescribeMessage {
+    kind: u8,	// 'S' to describe a prepared statement; or 'P' to describe a portal.
+    // we only support unnamed prepared stmt or portal
+}
+
+impl FeDescribeMessage {
+    pub fn parse(body: Bytes) -> Result<FeMessage> {
+        let mut buf = body.clone();
+        let kind = buf.get_u8();
+        let _pstmt_name = read_null_terminated(&mut buf)?;
+
+        // FIXME: see FeParseMessage::parse
+        /*
+        if pstmt_name.len() != 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "named prepared statements not implemented in Describe",
+            ));
+        }
+        */
+
+        if kind != 0x53 { // 'S'
```

Comment:
You might prefer: `if z == b'S'`

Reply:
A-ha, that's the syntax I was looking for, but my googling skills failed me. Thanks! :-)